packages/flutter/lib/src/material/stepper.dart

```diff
@@ -0,0 +1,577 @@
+  ///
+  /// If null, the subtitle is not shown.
+  final String subtitle;
+
+  /// The content of the step that appears below the title and subtitle.
+  ///
+  /// Underneath the content, every step as a 'continue' and 'cancel' button.
+  final Widget content;
+
+  /// The centered child of the step's colored circle.
+  ///
+  /// If null, the circle will be populated with the step's index within the
+  /// stepper.
+  final Widget circleChild;
+
+  /// The state of the step which determines the styling of its componenents
+  /// and whether steps are interactive.
+  final StepState state;
+}
+
+/// A Material stepper widget that keeps track of its current step and offers
+/// customizability and interactivity.
+///
+/// See also:
+///
+///  * [Step]
+///  * <https://material.google.com/components/steppers.html>
+class Stepper extends StatefulWidget {
+  /// Creates a stepper.
+  ///
+  /// Arguments [steps], [axis], and [currentStep] cannot be null.
+  Stepper({
+    Key key,
+    this.steps,
+    this.axis: Axis.vertical,
+    this.currentStep: 0,
+    this.onStepTapped,
+    this.onStepContinue,
+    this.onStepCancel
+  }) : super(key: key) {
+    assert(this.steps != null);
+    assert(this.axis != null);
+    assert(this.currentStep != null);
+  }
+
+  /// The steps of the stepper whose titles, subtitles, icons always get shown.
+  final List<Step> steps;
+
+  /// The axis on which the steps are layed out. In the case of
+  /// [Axis.horizontal], the content of the current step is displayed
+  /// underneath as opposed to the [Axis.vertical] case where it is displayed
+  /// in-between the steps.
+  final Axis axis;
+
+  /// The current step whose content is displayed.
+  final int currentStep;
+
+  /// The callback triggered when a step is tapped, with its index passed as
+  /// an argument.
+  final ValueChanged<int> onStepTapped;
+
+  /// The callback triggered when the 'continue' button is tapped.
+  ///
+  /// The index of the current step is not passed since the user is expected to
+  /// keep track of it in a parent state.
+  ///
+  /// If null, the buton will be disabled.
+  final VoidCallback onStepContinue;
+
+  /// The callback triggered when the 'cancel' button is tapped.
+  ///
+  /// The index of the current step is not passed since the user is expected to
+  /// keep track of it in a parent state.
+  ///
+  /// If null, the buton will be disabled.
+  final VoidCallback onStepCancel;
+
+  @override
+  _StepperState createState() => new _StepperState();
+}
+
+class _StepperState extends State<Stepper> {
+  List<GlobalKey> _keys;
+
+  @override
+  void initState() {
+    super.initState();
+    _keys = new List<GlobalKey>.generate(
+      config.steps.length,
+      (int i) => new GlobalKey()
+    );
+  }
+
+  @override
+  void didUpdateConfig(Stepper oldConfig) {
+    super.didUpdateConfig(oldConfig);
+    if (config.steps.length != oldConfig.steps.length)
+      _keys = new List<GlobalKey>.generate(
```

Comment:
I do think we should supporting adding steps dynamically, since the spec allows it explicitly. I would add a key to Step, and then I think we should require that either all Steps in a list have keys or none do, and then if none do then you imply ordinal keys for them, and if they do, then you use the provided keys. That handles adding/removing steps.